Dense vector field types updatable for int4

tteofili opened the pull request:
This PR makes it possible to update mappings also for the newly introduced int4 vectors (`int4_flat`, `int4_hnsw`), at the moment these are the allowed upgrade paths:
`flat` -> anything
`hnsw` -> `int8_hnsw`
`int8_flat` -> `hnsw`, `int8_hnsw`
`int8_hnsw` -> nothing
with int4 we may have the following upgrade paths:
`flat` -> anything
`hnsw` -> `int8_hnsw`, `int4_hnsw`
`int8_flat` -> `int4_flat`, `int8_hnsw`, `int4_hnsw`
`int4_flat` -> `hnsw`, `int8_hnsw`, `int4_hnsw`
`int8_hnsw` -> `int4_hnsw`
`int4_hnsw` -> nothing

tteofili committed:
Dense vector field types updatable  to int4

tteofili committed:
Test adjustments

tteofili committed:
Test adjustments

tteofili committed:
Test adjustments to avoid mixed clusters issues

tteofili committed:
Update docs/changelog/110928.yaml

elasticsearchmachine commented:
Hi @tteofili, I've created a changelog YAML for you.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:
```diff
@@ -1562,15 +1574,17 @@ boolean supportsElementType(ElementType elementType) {
                     || int8HnswIndexOptions.confidenceInterval != null
                         && confidenceInterval.equals(int8HnswIndexOptions.confidenceInterval);
+            } else {
+                updatable = update.type.equals(VectorIndexType.INT4_HNSW.name);
```
We don't care about a lower number of connections when updating int8 to int4?

benwtrent replied:
Agreed, the number of connections is important.

tteofili replied:
very good point, thanks!

mayya-sharipova commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/180_update_dense_vector_type.yml:
```diff
@@ -1320,6 +1470,104 @@ setup:
+      reason: 'updatable dense vector field type for int4 was added in 8.16'
+  - skip:
+      reason: "contains is a newly added assertion"
+      features: contains
```
I think the new way to speficy required runner feature is following:

```
  - requires:
      test_runner_features: [contains]
```

mayya-sharipova reviewed (approved):
@tteofili Thanks, overall LGTM! I've left a couple of comments.

Do we put somewhere in the documentation for users what types can be updated to what?

tteofili committed:
PR comments addressed

tteofili committed:
Merge branch 'in4_upgradable_dft' of github.com:tteofili/elasticsearch into in4_upgradable_dft

tteofili committed:
improved test coverage

tteofili commented:
thanks @benwtrent @mayya-sharipova for the reviews.
I think I've addressed your comments.

> Do we put somewhere in the documentation for users what types can be updated to what?
I am thinking to open a separate PR for the documentation bits.

tteofili committed:
Merge branch 'main' of github.com:elastic/elasticsearch into in4_upgradable_dft
